File: app/src/main/java/com/govnokoder/velotracker/BL/Controller/TrainingController.java
package com.govnokoder.velotracker.BL.Controller;

import android.content.Context;
import android.os.Environment;
import android.util.Log;

import com.google.gson.Gson;
import com.govnokoder.velotracker.BL.LineList;
import com.govnokoder.velotracker.BL.Model.Date;
import com.govnokoder.velotracker.BL.Model.Time;
import com.govnokoder.velotracker.BL.Model.Training;
import com.mapbox.mapboxsdk.geometry.LatLng;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;


public class TrainingController {
    private final String TRAINING_FILE_NAME = "data.json";
    private List<Training> trainings;
    private Training training;
    private Context context;

    public TrainingController(Context context){
        this.context = context;
        trainings = LoadTrainingsData();
        if(trainings == null)
        {
            trainings = new ArrayList<>();
        }
    }

    public void RemoveTraining(int index){
        trainings.remove(index);
        Save();
    }

    public void DeleteAll() {
        trainings.clear();
        Save();
    }

    public void setNewTrainingData(Date date, Time time, double distance,
                                   double maxSpeed, double averageSpeed, List<LineList> lines,
                                   LatLng startPoint, List<Long> heights, List<Double> speeds, long averageHeight, long maxHeight, long minHeight){
        training = new Training();
        training.Date = date;
        training.Time = time;
        training.Distance = distance;
        training.MaxSpeed = maxSpeed;
        training.AverageSpeed = averageSpeed;
        training.Lines = lines;
        training.setStartPoint(startPoint);
        training.MaxHeight = maxHeight;
        training.MinHeight = minHeight;
        training.AverageHeight = averageHeight;
        training.Heights = heights;
        training.Speeds = speeds;
        trainings.add(training);
        Save();
    }

    public List<Training>  LoadTrainingsData()
    {
        InputStreamReader streamReader = null;
        FileInputStream fileInputStream = null;
        try{
            fileInputStream = context.openFileInput(TRAINING_FILE_NAME);
            streamReader = new InputStreamReader(fileInputStream);
            Gson gson = new Gson();
            DataItems dataItems = gson.fromJson(streamReader, DataItems.class);
            return  dataItems.getTrainings();
        }
        catch (IOException ex){
            ex.printStackTrace();
        }
        finally {
            if (streamReader != null) {
                try {
                    streamReader.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (fileInputStream != null) {
                try {
                    fileInputStream.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return null;
    }

    private boolean Save()
    {
        Gson gson = new Gson();
        DataItems dataItems = new DataItems();
        dataItems.setTrainings(trainings);
        String jsonString = gson.toJson(dataItems);

        FileOutputStream fileOutputStream = null;

        try {
            fileOutputStream = context.openFileOutput(TRAINING_FILE_NAME, Context.MODE_PRIVATE);
            fileOutputStream.write(jsonString.getBytes());
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (fileOutputStream != null) {
                try {
                    fileOutputStream.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return false;
    }

    private static class DataItems {
        private List<Training> trainings;
        List<Training> getTrainings() {
            return trainings;
        }
        void setTrainings(List<Training> trainings) {
            this.trainings = trainings;
        }
    }
}

File: app/src/main/java/com/coursework/velotracker/BL/Controller/TrainingController.kt
package com.coursework.velotracker.BL.Controller

import android.content.Context
import com.coursework.velotracker.BL.Model.Training.TrainingRecorder
import com.coursework.velotracker.BL.Model.Training.TrainingStatistics
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStreamReader
import java.util.*
import kotlin.collections.ArrayList

class TrainingController(context: Context?) {

    private val TRAINING_FILE_NAME = "data.json"
    private var trainings: MutableList<TrainingStatistics> = ArrayList()
    private var currentTraining: TrainingStatistics = TrainingStatistics()
    private var context: Context? = context

    private var streamReader: InputStreamReader? = null
    private var fileInputStream: FileInputStream? = null
    private var fileOutputStream: FileOutputStream? = null

    fun removeTraining(index: Int) {
        trainings = loadTrainings()
        trainings.removeAt(index)
        save()
    }

    fun deleteAll() {
        trainings = loadTrainings()
        trainings.clear()
        save()
    }

    fun setNewTrainingData(trainingRecorder: TrainingRecorder) {
        currentTraining = TrainingStatistics(trainingRecorder)
        trainings = loadTrainings()
        trainings.add(trainings.size, currentTraining)
        save()
    }

    fun loadTrainings(): MutableList<TrainingStatistics> {
        fileInputStream = null
        streamReader = null
        try {
            return getTrainingsFromJson()
        } catch (ex: IOException) {
            ex.printStackTrace()
        } finally {
            closeStreamReader()
            closeFileInputStream()
        }
        return ArrayList()
    }

    private fun save(): Boolean {
        fileOutputStream = null
        try {
            saveTrainingsToGson()
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            closeFileOutputStream()
        }
        return false
    }

    private fun getTrainingsFromJson():MutableList<TrainingStatistics>{
        fileInputStream = context?.openFileInput(TRAINING_FILE_NAME)
        streamReader = InputStreamReader(fileInputStream)
        val gson = Gson()
        val dataItems: DataItems = gson.fromJson(streamReader, DataItems::class.java)
        return dataItems.trainings
    }

    private fun saveTrainingsToGson(){
        val gson = Gson()
        val dataItems = DataItems()
        dataItems.trainings = trainings
        val jsonString = gson.toJson(dataItems)
        val fileOutputStream = context?.openFileOutput(TRAINING_FILE_NAME, Context.MODE_PRIVATE)
        fileOutputStream?.write(jsonString.toByteArray())
    }

    private fun closeStreamReader(){
        try {
            streamReader?.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun closeFileInputStream(){
        try {
            fileInputStream?.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun closeFileOutputStream(){
        try {
            fileOutputStream?.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private class DataItems {
        var trainings: MutableList<TrainingStatistics> = ArrayList()
    }
}